src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -320,47 +316,42 @@ internal void MarkEntireType (TypeDefinition type, in DependencyInfo reason)
 			// In which case we would generate warnings with no source (hard to debug)
 			using var _ = _scopeStack.CurrentScope.Origin.MemberDefinition == null ? _scopeStack.PushScope (new MessageOrigin (type)) : null;
 
-			if (!_entireTypesMarked.Add (type))
-				return;
-
-			if (type.HasNestedTypes) {
-				foreach (TypeDefinition nested in type.NestedTypes)
-					MarkEntireType (nested, new DependencyInfo (DependencyKind.NestedType, type));
-			}
-
-			Annotations.Mark (type, reason);
-			MarkCustomAttributes (type, new DependencyInfo (DependencyKind.CustomAttribute, type));
-			MarkTypeSpecialCustomAttributes (type);
-
-			if (type.HasInterfaces) {
-				foreach (InterfaceImplementation iface in type.Interfaces)
-					MarkInterfaceImplementation (iface, new MessageOrigin (type));
+			// Calling MarkType here would result in extra warnings for cctors, ctors, delegate methods,
+			// serialization ctors, and overrides of preserved abstract methods.
+			// Instead, we only do extra type marking not already covered by the other cases in MarkEntireType.
+			void MarkTypeOnly (TypeDefinition type, in DependencyInfo reason)
+			{
+				Annotations.Mark (type, reason);
+				MarkCustomAttributes (type, new DependencyInfo (DependencyKind.CustomAttribute, type));
+				MarkTypeSpecialCustomAttributes (type);
+				MarkGenericParameterProvider (type);
 			}
 
-			MarkGenericParameterProvider (type);
+			MarkTypeOnly (type, reason);
```

Comment:
This comment is about the previous change in this area:
We don't do anything with base type here - it's the behavior we have for copy assemblies, so that's fine, but I think it's worth adding a comment - reading the code it looks really weird.

Side note: the base type will end up being marked in pretty much all cases anyway, since we call MarkMethod which will in turn call MarkType on the declaring type, which will mark the base type as appropriate.